### PR TITLE
added GaussianProposalDistribution

### PR DESCRIPTION
**Pull request by migwellian**
I withdrew the PR for a MultivariateGaussian proposal distribution, and instead submitted this one for the (univariate) Gaussian.

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistributionTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        assertThat(logProb, equalTo(expectedLogProb.sum()));
+    }
+
+    @Test
+    public void theLogProbAtFromIsGaussianAroundTheGivenPoint() {
+        double logProb = proposalDistribution.logProbAtFromGivenTo(proposal);
+        DoubleTensor expectedLogProb = Gaussian.withParameters(proposedState, sigma).logProb(currentState);
+        assertThat(logProb, equalTo(expectedLogProb.sum()));
+    }
+}
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistributionTest.java]: Nit: do we want a newline here?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistributionTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public void setRandomSeed() throws Exception {
+        KeanuRandom.setDefaultRandomSeed(0);
+    }
+
+    @Before
+    public void setUpMocks() throws Exception {
+        when(vertex1.getValue()).thenReturn(DoubleTensor.scalar(currentState.getValue(0)));
+        when(vertex2.getValue()).thenReturn(DoubleTensor.scalar(currentState.getValue(1)));
+
+        proposal = new Proposal();
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistributionTest.java]: Should the proposal setup be done in its own function (given the name of the function is to do with Mocks and this looks like just the test input).

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java [hunk 3]
```diff
@@ -167,6 +169,40 @@ public void samplesFromPriorWithObservedDeterministic() {
...
+            bayesNet,
+            Arrays.asList(A, B),
+            1000
+        );
+
+        double averagePosteriorA = posteriorSamples.getDoubleTensorSamples(A).getAverages().scalar();
+        double averagePosteriorB = posteriorSamples.getDoubleTensorSamples(B).getAverages().scalar();
+
+        double actual = averagePosteriorA + averagePosteriorB;
+        assertEquals(44.0, actual, 0.1);
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: Can you add a quick comment on how you derived the 44.0 value here please?

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistribution.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    public GaussianProposalDistribution(DoubleTensor sigma) {
+        this.sigma = sigma;
+    }
+
+    @Override
+    public Proposal getProposal(Set<Vertex> vertices, KeanuRandom random) {
+        Proposal proposal = new Proposal();
+        for (Vertex vertex : vertices) {
+            ContinuousDistribution proposalDistribution = Gaussian.withParameters((DoubleTensor) vertex.getValue(), sigma);
```
Comment (christophernorth) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistribution.java]: Casts are icky.  I wonder if we could fix up the class hierarchy somewhat to remove the need for this?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java (hunk 3 above)
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: um... ask @gordoncaleb? I copied it from the test above!

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistributionTest.java (hunk 1 above)
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistributionTest.java]: done

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistributionTest.java (hunk 2 above)
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistributionTest.java]: The order is important (I need to set up the mocks before the proposal) so they can't go in separate `@Before` methods.

I've renamed the method `setUpProposal` which is more accurate and hopefully addresses your point.

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistribution.java (hunk 4 above)
Comment (migwellian) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/GaussianProposalDistribution.java]: Agreed, casts are icky. The problem here (and many other places) is that we need a collection of vertices of mixed types. 

In fact I wonder if I ought to be filtering out those that are not `DoubleVertex` here. @gordoncaleb - can a proposal contain non-double-values vertices?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java (hunk 3 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: If you think about the result of the sum of A and B then you can see that the mu of C is also a gaussian of Gaussian(40, 2). Then if you take the log of their pdfs and sum them you get `(-2(x-40)^2)/8 - ((46-x)^2)/2`. If you find the root of the derivative of that (aka maximum) you will find it's 44. You can see this in WolfgramAlpha `D[(-2 (x - 40)^2)/8 - (46 - x)^2/2, x] == 0`. Maybe put this as a comment on the code?

**Pull request merged**